### ISSUE TITLE
PromQL: Dot in regexp should also match newlines (which is less surprising to most users and also allows optimising the `.* `regexp)

### ISSUE THREAD

**Issue by Betula-L** [prometheus/prometheus#7859]
**What did you do?**

I found that wildcard label matcher such as `app=~".*"` impact on query performance significantly.

Then i tried to edit benchmark `BenchmarkPostingsForMatchers` as followed,

![image](https://user-images.githubusercontent.com/6059935/91417449-628d2b00-e883-11ea-8855-bb24a9344e1d.png)

The results is amazing!!! 

The time consumption with wildcard is worser 100K times than without wildcard.  

```
BenchmarkPostingsForMatchers/Head/n="1",i=~".*",j="foo"-16                     5         213092008 ns/op
BenchmarkPostingsForMatchers/Head/n="1",j="foo"-16                        593829              1838 ns/op
```

**What did you expect to see?**

Label matcher `{n="1",i=~".*",j="foo}` is equal to `label matcher {n="1"j="foo}`, so the evaluation cost should be similar.

**Comment by roidelapluie** [comment]
Thank you.

This is equivalent in TSDB but not in PromQL, so the optimization could be done in tsdb. I think it would be great.

**Comment by roidelapluie** [comment]
Waiting for a tsdb maintainer to set P3 on this if they also find this useful.

**Comment by roidelapluie** [comment]
in promql:

- we need to send the matcher to remote reads
- `absent(foo{i=~".*",i="a"})` returns `{} 1` and `absent(foo{i="a"})` returns `{i="a"} 1`

**Comment by codesome** [comment]
Yeah it would be nice to have. We could also do it for `.+`.

**Comment by roidelapluie** [comment]
Didn't @pracucci made such optimizations already?

**Comment by roidelapluie** [comment]
I have found cef4dd6fff06198c61007211be460fb83cd708c1 and 2f6bf7de4c41015a9adeeb3e544766d61ec865eb.

**Comment by codesome** [comment]
That was for individual regex matchers

**Comment by Betula-L** [comment]
May i pick up this issue? 

I wonder that maybe i can take part into prometheus contribution.

**Comment by pracucci** [comment]
> This is equivalent in TSDB but not in PromQL, so the optimization could be done in tsdb. I think it would be great.

I think we tried to do this optimisation in the past, but have been told by @brian-brazil that it's not safe to just skip it because of label values with newlines.

**Comment by brian-brazil** [comment]
Yes, this is not a safe optimisation.

**Comment by roidelapluie** [comment]
I remember now indeed :)

**Comment by Betula-L** [comment]
It seems that we should use `!=""` rather use `.*` or `.+`  in the most situations now.

**Comment by pracucci** [comment]
> It seems that we should use `!=""` rather use `.*` or `.+` in the most situations now.

If you're currently using `app=~".*"` it also matches metrics without the `app` label at all. On the contrary, if you use `app!=""` it will not match metrics without the `app` label. The two label selectors are not the same thing.

**Comment by gouthamve** [comment]
Hrm, we've discussed this before and even sent a PR for this: https://github.com/prometheus/prometheus/pull/6996

This is not a safe optimisation with the current model, but something to consider for Prometheus 3.0. Will add this to the dev summit agenda and report back once we discuss this further!

**Comment by cyriltovena** [comment]
my 2 cents on this. 

There's a LOT of wasted CPU for this because a lot of people uses Grafana variables and when they want to match all they use `.*` or even `.+` I get the point of compatibility, but I'm sure no one is relying on `.*` to match all but newlines. 

I'm sure someone is using label values with newlines, does he expect to not matches his label when he does `.*` ? I'd be surprised. 

Honestly if we could count for how much money wasted in total across all companies in the world this does, I'm sure you'll change your mind.

**Comment by roidelapluie** [comment]
Another idea would be to change the prometheus default anchor behind a feature flag and optimize for (?s:.*) and (?s:.+) in tsdb as well, but the user would still type `=~".*"`.

**Comment by cyriltovena** [comment]
> ![allv](https://user-images.githubusercontent.com/291750/136232166-9551c253-c12e-40d2-9897-2e02c91d6fc1.png)
> 
>  It seems that grafana could change the "default auto" to be `(?s:.*)` and `(?s:.+)`, then we can optimize for these in tsdb.

For sure

**Comment by yeya24** [comment]
I think this is still worth tackling. @roidelapluie which way do we prefer more:
1. A feature flag to match `.*` and `.+` and handle them differently when matching postings?
2. A feature flag to change default anchor.

I feel option 1 is easier.

**Comment by roidelapluie** [comment]
But only option 2 is correct.

**Comment by yeya24** [comment]
> But only option 2 is correct.

Yes. Option 1 changes the behavior but it is something acceptable depending on the usecase. If no usecase to have new line in values then option 1 should be sufficient?

**Comment by bboreham** [comment]
I think ignoring newlines matches the expectation of most users of Prometheus.

**Comment by roidelapluie** [comment]
The plan was always correctness, so I don't see an issue with option 2. What would be wrong with it?

**Comment by alanprot** [comment]
At some point i was playing around with the `FastRegexMatcher` to opmize those regex: https://github.com/alanprot/prometheus/commit/bc168724872dddee212415502d26da72401528d7#diff-c6bfdf5ae2a5edbc3df844f48f297e4d2a4f4106662f1a04c8931fbb38f1fdae

Idk if this is something we wanna pursue.

**Comment by bwplotka** [comment]
Hello from the Bug Scrub!

> I think ignoring newlines matches the expectation of most users of Prometheus.

Sounds like [@gouthamve comment is still](https://github.com/prometheus/prometheus/issues/7859#issuecomment-683810012) relevant. Prometheus 3.0 coordinators, was this issue discussed already? We updated title as well.

It feels 9X.X% of users would likely be OK with this optimization, so let's decide what we want for Prometheus 3.0 (personally I think we should break it explicitly)

**Comment by beorn7** [comment]
Note the related #8525.

**Comment by jan--f** [comment]
@marioferh is working on this

**Comment by marioferh** [comment]
> .+

I've been trying to optimize also .+ but not sure how to do it. Any ideas?

**Comment by marioferh** [comment]
> Yeah it would be nice to have. We could also do it for `.+`.

I've been trying to optimize also .+ but not sure how to do it. Any ideas?

**Comment by bboreham** [comment]
Should be the same as `!=""` ?

**Comment by marioferh** [comment]
It's already the same and similar in performance:

```
		{
			matchers: []*labels.Matcher{labels.MustNewMatcher(labels.MatchEqual, "n", "1"), labels.MustNewMatcher(labels.MatchNotEqual, "i", "")},
			exp: []labels.Labels{
				labels.FromStrings("n", "1", "i", "a"),
				labels.FromStrings("n", "1", "i", "b"),
			},
		},
		{
			matchers: []*labels.Matcher{labels.MustNewMatcher(labels.MatchEqual, "n", "1"), labels.MustNewMatcher(labels.MatchRegexp, "i", "^.+$")},
			exp: []labels.Labels{
				labels.FromStrings("n", "1", "i", "a"),
				labels.FromStrings("n", "1", "i", "b"),
			},
		},

```
```

cpu: Intel(R) Core(TM) i7-10610U CPU @ 1.80GHz
BenchmarkQuerierMario/Block/PostingsForMatchers/i=~".+"-8             68          20133630 ns/op        18530016 B/op     100042 allocs/op
BenchmarkQuerierMario/Block/PostingsForMatchers/i!=""-8               56          19507532 ns/op        12817808 B/op     100016 allocs/op

```

**Comment by bboreham** [comment]
Closed by #14505 🎉
